lib/instrumentation/modules/@aws-sdk/client-dynamodb.js

```diff
@@ -0,0 +1,151 @@
+        } catch (ex) {
+          // Save the error for use in `finally` below, but re-throw it to
+          // not impact code flow.
+          err = ex;
+
+          // This code path happens with a GetObject conditional request
+          // that returns a 304 Not Modified.
+          statusCode = err && err.$metadata && err.$metadata.httpStatusCode;
+          throw ex;
+        } finally {
+          if (statusCode) {
+            span._setOutcomeFromHttpStatusCode(statusCode);
+          } else {
+            span._setOutcomeFromErrorCapture(constants.OUTCOME_FAILURE);
+          }
+          if (err && (!statusCode || statusCode >= 400)) {
+            agent.captureError(err, { skipOutcome: true });
+          }
+
+          // TODO: something related to region?
+          const config = client.config;
+          const region = await config.region();
+
+          // Set the db context
+          const dbContext = { type: SUBTYPE }; // dynamodb
+          if (region) {
+            dbContext.instance = region;
+          }
+          if (input && input.KeyConditionExpression) {
+            dbContext.statement = input.KeyConditionExpression;
+          }
+          span.setDbContext(dbContext);
+
+          // Set destination context
+          const destContext = {};
+          if (context[elasticAPMStash]) {
+            destContext.address = context[elasticAPMStash].hostname;
+            destContext.port = context[elasticAPMStash].port;
+          }
+          if (region) {
+            destContext.service = { resource: `dynamodb/${region}` };
+            destContext.cloud = { region };
+          }
+          span._setDestinationContext(destContext);
+
+          // TODO: review spec and add OTel attributes
+          // https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/semantic_conventions/trace/instrumentation/aws-sdk.yml#L435
+          // OTel attributes
+          // const otelAttrs = span._getOTelAttributes()
+
+          // otelAttrs['aws.dynamodb.system'] = 'dynamodb'
+          // otelAttrs['aws.dynamodb.operation'] = span.action
+          // if (table) {
+          //   otelAttrs['aws.dynamodb.table_names'] = table
+          // }
+          // if (input && input.Select) {
+          //   otelAttrs['aws.dynamodb.select'] = input.Select
+          // }
+          // if (input && input.ProjectionExpression) {
+          //   otelAttrs['aws.dynamodb.projection'] = input.ProjectionExpression
+          // }
+          // // TODO: check result for consumed_capacity
+          // if (result && result.ConsumedCapacity) {
+          //   otelAttrs['aws.dynamodb.consumed_capacity'] = JSON.stringify(result.ConsumedCapacity)
+          // }
+
+          span.end();
+        }
+
+        return result;
+      },
+      options: { step: 'initialize', priority: 'high', name: 'elasticAPMSpan' },
+    },
+    {
+      middleware: (next, context) => async (args) => {
```

Comment:
Note for the reviewer: this middleware has exactly the same implementation as the one in `client-s3`. My 1st reaction was to move to a separate file to reuse it but not sure if it may change in the future. Thoughts?

Reply:
I am fine either way. [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) is overrated.  Given we have only 3, it isn't a big deal either way. We can easily share them later if you prefer to not bother for now.

If we share them, then we'd have to share the `Symbol('elasticAPMStash')` as well because they are using different symbols currently.

Reply:
Let's keep it here :)